### ISSUE TITLE
:download --mhtml wastes diskspace with a lot of =09 and =20 escape sequences.

### ISSUE THREAD

**Issue by stroobandt** [The-Compiler/qutebrowser#2063]
`:download --mhtml` wastes diskspace with a lot of `=09` and `=20` escape sequences.
Seen in version `0.8.2` on Xubuntu 16.04 LTS.

**Comment by The-Compiler** [comment]
Those are tabs and spaces, and indeed according to wikipedia they only need to be encoded when at the end of a line:

> ASCII tab and space characters, decimal values 9 and 32, may be represented by themselves, except if these characters would appear at the end of the encoded line. In that case, they would need to be escaped as "`=09`" (tab) or "`=20`" (space), or be followed by a "`=`" (soft line break) as the last character of the encoded line. This last solution is valid because it prevents the tab or space from being the last character of the encoded line.

We simply use Python's `email.encoders.encode_quopri`, but it actually is quite simple - the relevant parts:

```python
from quopri import encodestring as _encodestring

def _qencode(s):
    enc = _encodestring(s, quotetabs=True)
    # Must encode spaces, which quopri.encodestring() doesn't do
    return enc.replace(b' ', b'=20')


def encode_quopri(msg):
    """Encode the message's payload in quoted-printable.

    Also, add an appropriate Content-Transfer-Encoding header.
    """
    orig = msg.get_payload(decode=True)
    encdata = _qencode(orig)
    msg.set_payload(encdata)
    msg['Content-Transfer-Encoding'] = 'quoted-printable'
```

So we could probably simply use `quopri.encodestring` without `quotetabs=True` (which *does* encode tabs/spaces at the end of line correctly), and do the rest of `encode_quopri` by hand.

cc @Kingdread

**Comment by Kingdread** [comment]
I'm pretty sure some of the first versions used `quopri.encodestring` directly, as seen here: https://github.com/The-Compiler/qutebrowser/commit/024ae52366cbbbdcbf88d4ab03bbc99e15e562ea

I mentioned some inconsistency, which iirc was this: https://bugs.python.org/issue20121, which is also why the `quopri.b2a_qp = quopri.a2b_qp = None` is in there (so it always uses the Python implementation). But I don't know anymore what exactly caused issues.

It seems strange though that it works with `email.encoders`, since that's just using `quopri` under the hood. Probably I just needed to adjust the function parameters :thinking:

**Comment by Kingdread** [comment]
Actually it seems like line endings are normalized by the email output generator, so the change should indeed be fairly easy.

**Comment by stroobandt** [comment]
As a reference, I have attached an MHTML file of <http://qutebrowser.org> generated by the UnMHT Firefox extension. See: [qutebrowser.unmht.mht](https://github.com/The-Compiler/qutebrowser/files/559893/qutebrowser.unmht.mht.txt)

**Comment by Raihid** [comment]
This was fixed by #2149
